File: CPPVersion/199/199.cpp
//
// Created by 郭蕴喆 on 2019/11/20.
//

#include "199.h"
#include <iostream>

vector<int> Solution::rightSideView(TreeNode *root) {
    if (root == NULL){
        vector<int> temp;
        return temp;
    }

    this->updateDic(root->val,0);
    int currentDepth=0;
    this->iterTree(root,currentDepth);
    int Depth=0;
    vector<int> ans;

    for (unordered_map<int,vector<int>>::const_iterator it=this->res.begin();it!=this->res.end();it++){
        if(it->first>=Depth){
            Depth=it->first;
        }
    }
    for(int i=0;i<=Depth;i++){
        ans.push_back(this->res[i].back());
    }
    return ans;
}

void Solution::updateDic(int A, int Length) {
    unordered_map<int,vector<int>>::const_iterator got=this->res.find(Length);
    if(got==this->res.end()){
        vector<int> temp(1,A);
        this->res.insert({Length,temp});
    } else{
        vector<int> temp=got->second;
        temp.push_back(A);
        this->res.erase(Length);
        this->res.insert({Length,temp});
    }
}

void Solution::iterTree(TreeNode *node, int currentDepth) {
    if(node->left != NULL){
        this->iterTree(node->left,currentDepth+1);
    }
    this->updateDic(node->val,currentDepth);
    if(node->right!=NULL){
        this->iterTree(node->right,currentDepth+1);
    }
}

void Solution::test(TreeNode *node) {
   vector<int> temp=this->rightSideView(node);
    for (vector<int>::const_iterator it=temp.begin();it!=temp.end();it++){
        cout<<' '<<*it;
    }
}

File: Medium/199/199.py
# Definition for a binary tree node.

class TreeNode:
    def __init__(self, x,Left,Right):
        self.val = x
        self.left = Left
        self.right = Right

class Solution:

    def updateDic(self,A:int,length:int):

        if self.res.__contains__(length):
            temp=self.res.get(length)
            temp.append(A)
            self.res.update({length:temp})
        else:
            self.res.setdefault(length,[A])

    def iterTree(self,node:TreeNode,currentDepth:int):
        if node.left is not None:
            self.iterTree(node.left,currentDepth+1)

        self.updateDic(node.val,currentDepth)

        if node.right is not None:
            self.iterTree(node.right,currentDepth+1)

    def rightSideView(self, root: TreeNode) -> [int]:
        if root is None:
            return []
        self.res={0:[root.val]}
        currentDepth=0
        self.iterTree(root,currentDepth)
        Depth=max(list(self.res.keys()))
        return [self.res.get(i)[-1] for i in range(0,Depth+1)]

if __name__ == '__main__':
    root=TreeNode(1,TreeNode(2,None,TreeNode(5,None,None)),TreeNode(3,None,TreeNode(4,TreeNode(7,None,None),TreeNode(8,None,None))))
    # root=TreeNode(1,TreeNode(2,TreeNode(4,None,None),None),TreeNode(3,None,None))
    S=Solution()
    print(S.rightSideView(root))
